packages/engine/paima-funnel/src/funnels/midnight/funnel.ts

```diff
@@ -0,0 +1,431 @@
+      )
+    );
+
+    for (let tx of block.transactions) {
+      for (let contractCall of tx.contractCalls) {
+        const extension = contractAddressToExtension.get(contractCall.address);
+        if (extension) {
+          // Only deserialize if we actually care.
+          const state = ContractState.deserialize(hexStringToUint8Array(contractCall.state));
+          // We could downcast contractCall to see if operations() contains something useful.
+          // For now let's report on the ledger variable contents.
+          const jsState = state.data.encode();
+          console.log(
+            '-- cde',
+            extension,
+            'contract @',
+            contractCall.address,
+            'has state',
+            JSON.stringify(jsState)
+          );
+
+          result.push({
+            blockNumber: baseBlockNumber,
+            transactionHash: tx.hash,
+
+            cdeDatumType: ChainDataExtensionDatumType.MidnightContractState,
+            cdeName: extension.cdeName,
+            network: this.chainName,
+
+            scheduledPrefix: extension.scheduledPrefix,
+            payload: JSON.stringify(jsState),
+          } satisfies CdeMidnightContractStateDatum);
+        } else {
+          console.log('discarding contract @', contractCall.address);
+        }
+      }
+    }
+
+    return result;
+  }
+
+  public override async readPresyncData(
+    args: ReadPresyncDataFrom
+  ): Promise<{ [network: number]: PresyncChainData[] | typeof FUNNEL_PRESYNC_FINISHED }> {
+    const baseDataPromise = this.baseFunnel.readPresyncData(args);
+
+    const startingBlock = await this.sharedData.mainNetworkApi.getStartingBlock();
+    if (!startingBlock) {
+      throw new Error("Couldn't get main's network staring block timestamp");
+    }
+
+    const start = this.cache.nextBlockHeight;
+    let block = await this.fetchBlock(this.cache.nextBlockHeight);
+    if (!(block && midnightTimestampToSeconds(block.timestamp) < startingBlock.timestamp)) {
+      // We're caught up. Return that we've finished.
+      const baseData = await baseDataPromise;
+      baseData[this.chainName] = FUNNEL_PRESYNC_FINISHED;
+      return baseData;
+    }
+
+    const result: PresyncChainData[] = [];
+
+    let lastLogTime = Date.now(),
+      lastLogBlock = start;
+    // TOOD: apply confirmation depth here
+    while (block && midnightTimestampToSeconds(block.timestamp) < startingBlock.timestamp) {
+      const now = Date.now();
+      if (now - lastLogTime > 5000) {
+        console.log(
+          `Midnight funnel ${this.config.networkId}: presync #${lastLogBlock}-${block.height}`
+        );
+
+        lastLogTime = now;
+        lastLogBlock = block.height + 1;
+      }
+
+      this.cache.nextBlockHeight = block.height + 1;
+
+      const extensionDatums = this.extensionsFromBlock(ENV.SM_START_BLOCKHEIGHT, block);
+      if (extensionDatums.length) {
+        result.push({
+          network: `midnight:${this.config.networkId}`,
```

Comment:
```suggestion
          caip2: `midnight:${this.config.networkId}`,
```

You'll want to use the change you made to `caip2PrefixFor` instead of hardcoding it here